File: fasten-core/bindings/swift/Sources/FastenStore/FastenStore.swift
import Foundation
import CFastenStoreCore

/// Thread-safe audit store backed by libfasten_core.
///
/// Build the Rust library first, then link it:
/// ```sh
/// cd fasten/fasten-core
/// cargo build --release --features all
/// swift build -Xlinker -L./target/release
/// ```
///
/// Usage:
/// ```swift
/// let store = try FastenStore(backend: "sqlite", connstr: ":memory:")
/// defer { store.close() }
/// try store.insert(["id": "evt-001", ...])
/// ```
public final class FastenStore {

    private var handle: OpaquePointer

    // ── Lifecycle ─────────────────────────────────────────────────────────────

    /// Open an audit store.
    ///
    /// - Parameters:
    ///   - backend: `"sqlite"` or `"postgres"`
    ///   - connstr: SQLite path / `":memory:"` or PostgreSQL DSN
    ///   - table:   plain or schema-qualified table name (default `"audit_log"`)
    public init(backend: String, connstr: String, table: String = "audit_log") throws {
        var errPtr: UnsafeMutablePointer<CChar>? = nil
        let h = fasten_store_open(backend, connstr, table, &errPtr)
        guard let h else {
            let msg = errPtr.map { String(cString: $0) } ?? "unknown error"
            errPtr.map { fasten_store_free_str($0) }
            throw FastenStoreError(msg)
        }
        handle = h
    }

    deinit { fasten_store_close(handle) }

    /// Release resources immediately.  The store is unusable after this call.
    /// Calling `close()` makes subsequent operations throw.
    public func close() {
        fasten_store_close(handle)
        // Note: after close the handle is dangling; subsequent calls will crash.
        // Callers should discard the reference after close().
    }

    // ── Write path ────────────────────────────────────────────────────────────

    /// Insert one audit row.  Accepts any `Encodable` value.
    /// Duplicate IDs are silently ignored.
    public func insert(_ row: some Encodable) throws {
        let data = try JSONEncoder().encode(row)
        guard let json = String(data: data, encoding: .utf8) else {
            throw FastenStoreError("row could not be encoded as UTF-8 JSON")
        }
        try insertJSON(json)
    }

    /// Insert from a plain dictionary (useful in scripts / tests).
    public func insert(_ row: [String: Any]) throws {
        let data = try JSONSerialization.data(withJSONObject: row)
        guard let json = String(data: data, encoding: .utf8) else {
            throw FastenStoreError("row could not be encoded as UTF-8 JSON")
        }
        try insertJSON(json)
    }

    /// Insert from a pre-serialised JSON string.
    public func insertJSON(_ json: String) throws {
        var errPtr: UnsafeMutablePointer<CChar>? = nil
        let rc = fasten_store_insert(handle, json, &errPtr)
        guard rc == 0 else {
            let msg = errPtr.map { String(cString: $0) } ?? "insert failed"
            errPtr.map { fasten_store_free_str($0) }
            throw FastenStoreError(msg)
        }
    }

    // ── Health ────────────────────────────────────────────────────────────────

    /// Verify the backend is reachable (runs `SELECT 1`).
    public func ping() throws {
        var errPtr: UnsafeMutablePointer<CChar>? = nil
        let rc = fasten_store_ping(handle, &errPtr)
        guard rc == 0 else {
            let msg = errPtr.map { String(cString: $0) } ?? "ping failed"
            errPtr.map { fasten_store_free_str($0) }
            throw FastenStoreError(msg)
        }
    }

    // ── Metadata ──────────────────────────────────────────────────────────────

    /// Return the library version string, e.g. `"0.1.0"`.
    public static var version: String {
        String(cString: fasten_store_version())
    }
}

File: fasten-core/bindings/java/src/main/java/sh/fasten/store/FastenStore.java
package sh.fasten.store;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.sun.jna.Pointer;
import com.sun.jna.ptr.PointerByReference;

import java.io.IOException;
import java.util.Map;

/**
 * Thread-safe Java client for an audit store backed by libfasten_store_core.
 *
 * <p>Implements {@link AutoCloseable} for use with try-with-resources:
 * <pre>{@code
 * try (var store = FastenStore.open("sqlite", ":memory:", "audit_log")) {
 *     store.insert(rowMap);
 * }
 * }</pre>
 *
 * <p>Library loading: set {@code FASTEN_STORE_CORE_LIB} to the absolute path of
 * the .so/.dylib/.dll, or ensure {@code libfasten_store_core} is on
 * {@code LD_LIBRARY_PATH} / {@code DYLD_LIBRARY_PATH} / the system library path.
 */
public final class FastenStore implements AutoCloseable {

    private static final FastenStoreLib LIB = FastenStoreLib.INSTANCE;
    private static final ObjectMapper   JSON = new ObjectMapper();

    private volatile Pointer handle;

    private FastenStore(Pointer handle) {
        this.handle = handle;
    }

    // ── Factory ───────────────────────────────────────────────────────────────

    /**
     * Open an audit store.
     *
     * @param backend  {@code "sqlite"} or {@code "postgres"}
     * @param connstr  SQLite path / {@code ":memory:"} or PostgreSQL DSN
     * @param table    plain or schema-qualified table name
     * @throws FastenStoreException if the backend cannot be opened
     */
    public static FastenStore open(String backend, String connstr, String table) {
        var errRef = new PointerByReference(Pointer.NULL);
        Pointer h = LIB.fasten_store_open(backend, connstr, table, errRef);
        if (h == null) {
            throw new FastenStoreException(takeError(errRef));
        }
        return new FastenStore(h);
    }

    /** Convenience overload — uses {@code "audit_log"} as the table name. */
    public static FastenStore open(String backend, String connstr) {
        return open(backend, connstr, "audit_log");
    }

    // ── Write path ────────────────────────────────────────────────────────────

    /**
     * Insert one audit row.  Accepts any {@link Map} or POJO that Jackson
     * can serialise.  Duplicate IDs are silently ignored.
     *
     * @throws FastenStoreException on backend error
     */
    public void insert(Object row) {
        String json;
        try {
            json = JSON.writeValueAsString(row);
        } catch (IOException e) {
            throw new FastenStoreException("JSON serialisation failed", e);
        }
        insertJson(json);
    }

    /**
     * Insert from a pre-serialised JSON string.
     *
     * @throws FastenStoreException on backend error
     */
    public void insertJson(String rowJson) {
        var errRef = new PointerByReference(Pointer.NULL);
        int rc = LIB.fasten_store_insert(handle, rowJson, errRef);
        if (rc != 0) {
            throw new FastenStoreException(takeError(errRef));
        }
    }

    // ── Health ────────────────────────────────────────────────────────────────

    /**
     * Verify the backend is reachable (runs {@code SELECT 1}).
     *
     * @throws FastenStoreException if unreachable
     */
    public void ping() {
        var errRef = new PointerByReference(Pointer.NULL);
        int rc = LIB.fasten_store_ping(handle, errRef);
        if (rc != 0) {
            throw new FastenStoreException(takeError(errRef));
        }
    }

    // ── Lifecycle ─────────────────────────────────────────────────────────────

    /** Release all resources.  Safe to call multiple times. */
    @Override
    public synchronized void close() {
        if (handle != null) {
            LIB.fasten_store_close(handle);
            handle = null;
        }
    }

    // ── Metadata ──────────────────────────────────────────────────────────────

    /** Return the library version string, e.g. {@code "0.1.0"}. */
    public static String version() {
        return LIB.fasten_store_version();
    }

    // ── Internal ──────────────────────────────────────────────────────────────

    private static String takeError(PointerByReference errRef) {
        Pointer p = errRef.getValue();
        if (p == null || p.equals(Pointer.NULL)) {
            return "(no detail)";
        }
        String msg = p.getString(0);
        LIB.fasten_store_free_str(p);
        return msg;
    }
}
